### PR TITLE
Introduce a sentinel value _NO_VALUE to improve Global resolvers to support defaults `0` or `None`

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

Allow passing `None` values as default for globals

## Development notes

There's two things reported by a user -
1) There was a bug which wouldn't allow a zero value to be set for globals - example `globals.yml`
    ```
    maxLookBackDays: 0  --> Doesnt work
    maxLookBackDays: 1  --> It works
    ```
2) `None` default value support

> If we can get a null/None support that would also be nice as i wanted to pass null as default in some cases but it wasn't possible

We do allow parameters to be set as None, so I feel like this was reasonable. I've added a logger warning if a `None` default value is being used for a global variable used somewhere.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/config/omegaconf_config.py
Comment (noklam) [inline kedro/config/omegaconf_config.py:343]: Why do we care about the `value` here? It could be anything included `None` or `0`. Shouldn't we check where the `key` is a valid key instead of the value?

**Comment by noklam** [comment]
We should fix this before we release. cc@astrojuanlu

**Comment by astrojuanlu** [comment]
I don't fully understand the context of this one. I get this part

```
maxLookBackDays: 0  --> Doesnt work
maxLookBackDays: 1  --> It works
```

but not part (2) or your comment @noklam , could you elaborate?

**Comment by ankatiyar** [comment]
@astrojuanlu First part is a bug which definitely needs fixing. The second part was that if someone has set their default value to `None`, it shouldn't error out and just set the value of the parameter to `None` - 
```
my_param: "${globals:whatever, None}"
```

**Comment by noklam** [comment]
@astrojuanlu My comment is more on the implementation side, the bug is a result of missing conditions on "value". My point is we shouldn't validate the value (It can be 0 or None, it's a bug now because we use `not`), we should check if the "key" exists and assign the value as is.

File: kedro/config/omegaconf_config.py
Comment (ankatiyar) [reply kedro/config/omegaconf_config.py]: I've made some changes!

**Review by merelcht — APPROVED** [review]
LGTM! 👍 I agree this should go into the release.

**Review by SajidAlamQB — APPROVED** [review]

**Comment by noklam** [comment]
I saw it has 2 approval already, but I think there are still bugs. I am working on a draft to fix this.

**Comment by noklam** [comment]
Cc @SajidAlamQB #2983

**Comment by noklam** [comment]
I think we should have a sentinel value `NO_VALUE` rather than `None` to differentiate a user define default `None` versus a no-op

1. `${globals: DUMMY_KEY, None}`
2.`${globals: DUMMY_KEY}`

Right now we cannot differentiate between the two. When the `globals` is missing,  2. should raise an error while 1 should not (

I can push the new change soon.

**Comment by noklam** [comment]
Found more bug and I am adding more tests to catch all the error.
1. [c80437f](https://github.com/kedro-org/kedro/pull/2976/commits/c80437fcd53c6ef547f2b6925dd949ab4357b069) - for handling non-existing key
2. [b36ffbd](https://github.com/kedro-org/kedro/pull/2976/commits/b36ffbdb6ec1374efc8692986d78b512d59914cf) - to handle a special case when the `globals` sub-key is not a dictionary.

I wonder can we handle this better with `omegaconf`, similar to the way we parse `--params` because we are running into the same traps that we need to handle `dict` and traversing the nested key  ourselves.

**Review by merelcht — COMMENTED** [review]
Thanks for digging a bit deeper in this and finding all edge cases @noklam ⭐ It looks like it needs some more test coverage. I'm happy to approve after that.

**Comment by ankatiyar** [comment]
@merelcht @noklam I was actually testing this but it has some issues too - 
If your `globals.yml` has this -
```yaml
x: null
```
And in your `parameters.yml` -
```yaml
param : "${globals:x}" # This will throw an interpolation error
param2: "${globals:x, 56}" # This will set param2 = 56
```
This is because the `Omegaconf.select()` fn returns `None` if the key doesn't exist or it actually is `null`, in which case it uses the default.
What do you think about reverting to the previous implementation but keeping the sentinel value?
wrt: 

> [b36ffbd](https://github.com/kedro-org/kedro/pull/2976/commits/b36ffbdb6ec1374efc8692986d78b512d59914cf) - to handle a special case when the globals sub-key is not a dictionary.

This I think is an edge case we probably won't come across often.

**Review by SajidAlamQB — APPROVED** [review]
I am happy with this very nice work @ankatiyar and @noklam! 🌟

File: kedro/config/omegaconf_config.py
Comment (noklam) [reply kedro/config/omegaconf_config.py:339]: Yes. Currently there are 3 paths it can take:
1. exists in `globals.yml` (take the global value as is)
2. Not define in `globals.yml`, but default provided
3. Not define in `globals.yml` and not default provided - raise `InterpolationResolutionError`

Comment (ankatiyar) [reply kedro/config/omegaconf_config.py:339]: Done

**Pull request merged**
